Parallelize execution of unit tests in hack/test-go.sh.

ixdy opened the pull request:
#4574 slowed down hack/test-go.sh considerably, especially when code coverage was not requested. This should restore the previous speedy behavior for non-coverage runs, as well as speed up coverage runs.

Fixes #4713.

ixdy committed:
Renable code coverage collection in unit tests, including on Travis.

When code coverage is not being collected, just issue a single 'go test'
command, as is already done. Go will internally parallize execution.

When code coverage is being collected, it is necessary to issue separate
'go test' commands for each package, since Go does not support
collecting coverage across packages. Using xargs -P will parallelize
these invocations, however, speeding up test execution.
The number of simultaneous processes to use can be specified with
KUBE_COVERPROCS.

Update Travis config to pass along the number of CPUs to use for running
tests.

satnam6502 commented:
@filbranden : do you have time to give this a look?

thockin commented on hack/test-go.sh:
```diff
@@ -136,56 +138,58 @@ if [[ $iterations -gt 1 ]]; then
+# command, which is much faster.
+if [[ -z "${KUBE_COVER}" ]]; then
+  kube::log::status "Running unit tests without code coverage"
+  echo "${test_dirs}" | xargs -n1 printf "${KUBE_GO_PACKAGE}/%s\n" | \
```
See the printf-for-arrays trick in #4714 - I think it's more elegant, but I defer to you guys as to best final answer.

filbranden replied:
test_dirs is wrong either way. It should have been an array. And yes use `${array[@]/#/${KUBE_GO_PACKAGE}/}` to prefix it with something else.

filbranden commented on hack/test-go.sh:
```diff
@@ -45,6 +45,8 @@ KUBE_TIMEOUT=${KUBE_TIMEOUT:--timeout 120s}
 KUBE_COVERMODE=${KUBE_COVERMODE:-atomic}
 KUBE_RACE=${KUBE_RACE:-}   # use KUBE_RACE="-race" to enable race testing
+# How many processes to use when running tests.
+KUBE_PROCS=${KUBE_PROCS:-$(grep -c '^processor' /proc/cpuinfo)}
```
Looks flaky to me.

Will probably annoy people building it on MacOS and I know of quite a few of them.

Isn't this actually the default in Go? Can't it just take a parameter to do the auto-detection by itself?

filbranden commented on hack/test-go.sh:
```diff
@@ -136,56 +138,58 @@ if [[ $iterations -gt 1 ]]; then
+if [[ -z "${KUBE_COVER}" ]]; then
+  kube::log::status "Running unit tests without code coverage"
+  echo "${test_dirs}" | xargs -n1 printf "${KUBE_GO_PACKAGE}/%s\n" | \
+      xargs go test -p=${KUBE_PROCS} "${goflags[@]:+${goflags[@]}}" \
```
`xargs` is probably not helping here since you end up with a single command line anyways... Might as well just expand the array here (it's unlikely it will exceed the maximum number of arguments that the shell can take.)

ixdy commented:
Thanks for the review - I've addressed your comments here + the ones on #4714.
I'm not entirely sure whether the parameters I've supplied (KUBE_COVERPROCS=4 with no -p to go) will work on Travis, so if that fails I may still need to tune.

Otherwise, can you please take another look? Thanks!

filbranden commented on hack/test-go.sh:
```diff
@@ -101,6 +104,11 @@ for arg; do
   fi
 done
+# If either no test cases were provided, or if the first test case is an empty
+# string (needed to specify Go arguments?), run all tests.
```
The empty string case doesn't make sense to me. Can you give me an example of why it would be needed?

ixdy replied:
a few places were using "" as the first argument to bypass getopt intercepting Go args.
I updated the callers to use -- instead and removed this special case.

filbranden commented on hack/test-go.sh:
```diff
@@ -136,56 +144,51 @@ if [[ $iterations -gt 1 ]]; then
+# Include all coverage reach data in the combined profile, but exclude the
+# 'mode' lines, as there should be only one.
+find "${cover_report_dir}" -name "${cover_profile}" | \
+  xargs -I{} grep -h -v "^mode:" {} >>"${combined_cover_profile}" || true
```
Use a single xargs grep, no reason to do file by file.

Also, a single redirect:

```shell
combined_cover_profile="${cover_report_dir}/combined-coverage.out"
{
  echo "mode: ${KUBE_COVERMODE}"
  find "${cover_report_dir}" -name "${cover_profile}" | xargs -r grep -h -v "^mode:"
} >"${combined_cover_profile}"
```

ixdy commented:
Still tuning Travis params.

ixdy commented:
There's quite a bit of variability in test runtime on Travis, but it looks like this change now adds around 15-20s to the Travis runtime (vs not reporting coverage), which is hopefully more acceptable.

Please take another look when you get a chance? Thanks.

thockin commented:
FWIW I don't care too much how long Travis takes, but the diff between :45
and 2:00 on every 'make' is infuriating.  Why the hell is go re-running
tests for things it didn't have to recompile anyway?  Can we fix THAT
instead?

On Mon, Feb 23, 2015 at 4:49 PM, Jeff Grafton <notifications@github.com>
wrote:

> There's quite a bit of variability in test runtime on Travis, but it looks
> like this change now adds around 15-20s to the Travis runtime (vs not
> reporting coverage), which is hopefully more acceptable.
>
> Please take another look when you get a chance? Thanks.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4686#issuecomment-75675990>
> .
>

ixdy commented:
@thockin the "make test" version should be the same now before and after this change; it is a single 'go test' call, rather than N 'go test' calls as was in #4574.

When coverage is enabled it is still N 'go test' calls (this will affect Travis), but it should be much faster than #4574 was.

thockin commented:
My question was whether there is a way to tell "go test" to just run tests
that have change since the last run.

On Mon, Feb 23, 2015 at 7:40 PM, Jeff Grafton <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> the "make test" version should be
> the same now before and after this change; it is a single 'go test' call,
> rather than N 'go test' calls as was in #4574
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4574>.
>
> When coverage is enabled it is still N 'go test' calls (this will affect
> Travis), but it should be much faster than #4574
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4574> was.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4686#issuecomment-75694260>
> .
>

ixdy commented:
@thockin I don't think "go test" has any sense of caching test results, so I'm not sure it's possible to make it run only unaffected tests. in any case, it seems like that is out of scope for this PR.
@filbranden ping?

thockin commented:
Why does Go hate me so much?

On Fri, Feb 27, 2015 at 12:32 PM, Jeff Grafton <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> I don't think "go test" has any
> sense of caching test results, so I'm not sure it's possible to make it run
> only unaffected tests. in any case, it seems like that is out of scope for
> this PR.
> @filbranden <https://github.com/filbranden> ping?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4686#issuecomment-76466565>
> .
>

ixdy commented:
I'm curious where you get the ":45 to 2:00" figure. 'make test' consistently runs in ~21s on my machine.

thockin commented:
That was my laptop - runnign a single 'go test' that did all 100-ish dirs
was :45 and running them in serial was 2:00+

On Fri, Feb 27, 2015 at 2:35 PM, Jeff Grafton <notifications@github.com>
wrote:

> I'm curious where you get the ":45 to 2:00" figure. 'make test'
> consistently runs in ~21s on my machine.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4686#issuecomment-76485659>
> .
>

filbranden commented:
@thockin Are you still seeing a slowdown on this one or is it good to submit now?

thockin commented:
What I see today is pretty good, I think - sub 1 minute

On Mon, Mar 2, 2015 at 1:40 PM, Filipe Brandenburger <
notifications@github.com> wrote:

> @thockin <https://github.com/thockin> Are you still seeing a slowdown on
> this one or is it good to submit now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4686#issuecomment-76831413>
> .
>

filbranden commented:
@thockin Let us know if you have a slowdown after the merge...

Cheers,
Filipe

lavalamp commented:
Why does passing a "y" to KUBE_COVER work? When I run that locally, it is unhappy. Trying to understand why sometimes our builds fail in shippable with:

```
KUBE_RACE="-race" KUBE_COVER="-cover -covermode=atomic" KUBE_TIMEOUT='-timeout 60s' ./hack/test-go.sh "" -p=4
+++ Running unit tests without code coverage
can't load package: package github.com/GoogleCloudPlatform/kubernetes: no buildable Go source files in /home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes
!!! Error in ./hack/test-go.sh:149
  'go test "${goflags[@]:+${goflags[@]}}" ${KUBE_RACE} ${KUBE_TIMEOUT} "${@+${@/#/${KUBE_GO_PACKAGE}/}}"' exited with status 1
Call stack:
  1: ./hack/test-go.sh:149 main(...)
Exiting with status 1
```

ixdy commented:
@lavalamp what build is that from? the arguments there (KUBE_COVER="-cover ...") are old.

Combined with the "can't load package: package" error it looks unrelated to this PR.

lavalamp commented:
@ixdy Yeah I don't think that it's related to this PR, I just noticed that it's not using the line in our .travis.yml, tried to run the line in that file locally, and got a similar error:

```
$ KUBE_RACE="-race" KUBE_COVER="y" KUBE_GOVERALLS_BIN="$HOME/gopath/bin/goveralls" KUBE_TIMEOUT='-timeout 60s' KUBE_COVERPROCS=8 ./hack/test-go.sh -- -p=2
can't load package: package y: cannot find package "y" in any of:
        /usr/local/go/src/y (from $GOROOT)
        /usr/local/google/home/dbsmith/code/k/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/y (from $GOPATH)
        /usr/local/google/home/dbsmith/code/k/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/y
!!! Error in ./hack/test-go.sh:164
  'xargs go test "${goflags[@]:+${goflags[@]}}" ${KUBE_RACE} ${KUBE_TIMEOUT} ${KUBE_COVER}' exited with status 0 123
Call stack:
  1: ./hack/test-go.sh:164 main(...)
Exiting with status 1
```

ixdy commented:
Is your repo up to date? It looks like you have an old version of hack/test-go.sh.

lavalamp commented:
Thanks, indeed I was not running this in the branch I thought I was. Sorry for the noise!
